compat/getaddrinfo.cc

```diff
@@ -136,7 +130,7 @@ xgetaddrinfo (const char *nodename, const char *servname,
         /* Note: maintain port in host byte order to make debugging easier */
         if (isdigit (*servname))
             port = strtol (servname, nullptr, 10);
-        else if ((servent = getservbyname (servname, socktype)))
+        else if ((const struct servent = getservbyname (servname, socktype)))
```

Comment:
Wrong type and you can actually simplify this further:

```suggestion
        else if (const auto servent = getservbyname(servname, socktype))
```

Reply:
Small correction. We should always include the ```*``` qualifier to ensure robust type deduction when the variable is a raw pointer (same for references).

Reply:
> Small correction. We should always include the `*` qualifier 

We do not have a policy about it. IMHO, it is best to omit `*` in code that does not care whether the type is a raw pointer or a smart pointer (or a pointer at all). This change request is attached to an (imperfect) example of such code -- code that does not and should not care about such details. I do not recommend adding `*` to my suggested correction.

In his Almost Always Auto (AAA) [post](https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/), Herb Sutter considers asterisk-free auto declarations to be "modern C++ style" and gives many arguments in support of that style in general (while acknowledging that it requires some learning/adjustment/practice). Many disagree, of course. In my experience, most anti-auto (and for-asterisk) arguments are rooted in unrelated problems. Auto just amplifies those problems but gets blamed for them.

When used correctly, an explicit `*` does not change the resulting type (i.e. there is no type-based advantage of using it), but an explicit `*` is not compatible with smart pointers (and it obviously uses more ink). In a sense, it is not an explicit `*` that makes a variable a pointer -- it is how that variable is used (or the set of operations the variable supports).

It is also not trivial to use `*` correctly -- there are many [ways](https://godbolt.org/z/aTP5GKG1e) to add an explicit `*`:

```C++
const auto y1 = x; // a constant pointer; whether it is a pointer to a constant depends on whether x is a pointer to a constant
auto * const y2 = x; // the same thing said using more ink

const auto *y3 = x; // a non-constant pointer to a constant, even if x is not a pointer to a constant

const auto * const y4 = getservbyname(); // a constant pointer to a constant, even if x is not a pointer to a constant
```


> to ensure robust type deduction when the variable is a raw pointer

An explicit `*` does not make type deduction "robust" (for any definition of "robust" that I can think of). In fact, it is kind of the opposite -- an explicit `*` makes type deduction more susceptible to source type changes (i.e. makes the declaration _less_ "robust" in the presence of code changes).

Using an explicit `*` means that the author either needs to add memory constness (y3 and y4 above) or needs to insist that the pointer is a raw pointer. In most use cases, there is no such need.

With all other factors being equal, we should aim for code that does not need to know whether the variable is a _raw_ pointer. That aspect may change and most correctly written code should not be affected by such changes.

> (same for references).

References are a completely different use case! With references, an explicit `&` is _required_ by C++. One simply cannot declare an auto-typed _reference_ without an explicit `&`.

Reply:
I'd follow Alex' idea and let the compiler deal with it.
With the recent move to c++17, we might want to start using unique_ptr and shared_ptr more unless there's any reason why our refcount implementation is better - is there one?

Reply:
To avoid misunderstanding, the discussion below is unrelated to the primary/original concern of this change request: `auto` is a better default than `auto *` _regardless_ of specific smart pointer types in use.

----

> we might want to start using shared_ptr more unless there's any reason why our refcount implementation is better - is there one?

RefCount is an _invasive_ shared pointer. Compared to non-invasive std::shared_ptr, RefCount advantages include:

* The ability to obtain the right shared pointer from a raw object pointer. A lot of current Squid code relies on this feature IIRC.
* No extra memory allocation (to hold an internal "control block" object containing reference counting metadata that is shared by all users of a shared pointer instance) in all use cases. std::shared_ptr prevents that extra memory allocation only when created by calling [std::make_shared](https://en.cppreference.com/w/cpp/memory/shared_ptr/make_shared) or [std::allocate_shared](https://en.cppreference.com/w/cpp/memory/shared_ptr/allocate_shared), and that kind of creation is kind of going against our current Mem::Allocator architecture by allocating objects of different sizes/types than the (often pooled) object returned by dereferencing std::shared_ptr.

In summary, there should be no mechanical conversion to std::shared_ptr. Each conversion case must be carefully evaluated. In some cases, the correct conversion target should not be std::shared_ptr but std::unique_ptr or no pointer at all! If we start using std::shared_ptr, we may need to add proper tools to integrate it with memory pools.

None of the above concerns apply to std::unique_ptr (which has a very different purpose compared to any shared pointer). That pointer should be used as needed.